Implement flexible PubSub API

jwilm opened the pull request:
Originally, pubsub support as only available only by creating a client
and then asking for a pubsub API. The original `PubSub` type owned a
connection. This neatly avoided the issue that only certain commands are
available once calling subscribe or psubscribe; however, it's very
unfriendly to connection poolers such as r2d2

This commit makes it so that a borrowed connection may be used for
PubSub. Available commands are limited while in pubsub mode by borring
the connection mutably into a type that only exposes the allowed
commands. This API may be accessed by calling `as_pubsub` on a
`Connection`. To go back to being able to use *any* command, this
wrapper type may be dropped, and the connection should become usable
again for other purposes.

Additionally, convenience methods are added to `Connection`: `subscribe`
and `psubscribe`. They accept a `ToRedisArgs` and a closure which will
be called every time a message is received. The subscription status can
be managed via the return value of the closure which is either
`ControlFlow::Continue` or `ControlFlow::Break`.

This is a breaking change; the original PubSub type was removed in favor
of this more flexible API and to avoid code bloat by having multiple
implementations or additional abstractions to allow an implementation to
be shared.

Resolves #135.

jwilm commented:
Test failures were "Connection reset by peer" errors. Maybe just needs to be rerun?

dbettin commented:
Is this close to being merged? If not, anything specifically blocking it?

badboy commented on src/commands.rs:
```diff
@@ -734,9 +734,100 @@ implement_commands! {
+pub trait PubSubCommands {
+    /// Subscribe to a list of channels using SUBSCRIBE and run the provided
+    /// closure for each message received.
+    fn subscribe<'a, C, F, U>(&mut self, _: C, _: F) -> RedisResult<U>
```
It might be good to explicitly state here that this is blocking until `Break` is returned. What do you think?

jwilm replied:
Good idea, I'll add it.

badboy commented on src/commands.rs:
```diff
@@ -734,9 +734,100 @@ implement_commands! {
+              C: ToRedisArgs;
+
+    /// Subscribe to a list of channels using PSUBSCRIBE and run the provided
+    /// closure for each message received.
```
As above.

badboy commented on src/connection.rs:
```diff
@@ -512,6 +492,13 @@ impl PubSub {
+impl<'a> Drop for PubSub<'a> {
+    fn drop(&mut self) {
+        let _ = cmd("UNSUBSCRIBE").query::<()>(self.con);
+        let _ = cmd("PUNSUBSCRIBE").query::<()>(self.con);
```
Unsubscribing will trigger a message per channel that's unsubscribed.
Will this cause problems for the underlying connection if it is later reused?
Having a test for this would be great.

jwilm replied:
It shouldn't because `query` will already expect to receive a result from the server and parse it. I agree that we should have a test for this including that the connection is no longer subscribed to anything.

jwilm replied:
As it turns out, a message per channel actually means a distinct message per channel at a protocol level. This will take a bit of work to support. Nice catch!

badboy commented on src/connection.rs:
```diff
@@ -427,44 +423,28 @@ impl ConnectionLike for Connection {
-        let chan = self.get_channel(&channel);
-        let _: () = try!(cmd("SUBSCRIBE").arg(&*chan).query(&self.con));
-        self.channels.insert(chan);
+        let _: () = try!(cmd("SUBSCRIBE").arg(channel).query(self.con));
```
While we're at it, can we switch to the postfix question mark operator here (and below)?

jwilm replied:
Yep, happy to do so. I believe I went with `try!` to be consistent with the rest of the code base.

jwilm commented:
@badboy I've addressed the issues you pointed to in your review. Please see the description on the final commit for details.

One point worth reiterating is that the updated PubSub API now consumes the connection, and it's only returned upon a successful call to `PubSub::into_inner`. There are two drawbacks to this:

1. Connections from a connection pooler won't be able to utilize the PubSub API
2. A failed call to `PubSub::into_inner` will cause the connection to be dropped.

On the flip side, this allows us to guarantee that the connection will be in a usable state when going in and out of PubSub state.

jwilm commented:
Reflecting a bit on the updated implementation, the ownership requirements imposed are too extreme, and they are contrary to the goals set out at the start of this PR. However, I have another idea about how to implement this.

Instead of having PubSub take ownership of connection, we can return to taking a mutable reference. The connection would itself gain a flag that indicates whether it is in the pubsub state. When entering this state, the flag would be raised. On an unsuccessful exit from this state, the flag would remain. Subsequent commands could check this flag and attempt to exit the pubsub state, or else return an appropriate error.

This solution seems to be the best of both worlds -- good ergonomics, connection pooling support, and pubsub constraints can be enforced while making it clear that a connection is not usable by returning an appropriate error on subsequent usage if exiting the pubsub state fails.

What do you think?

badboy commented:
@jwilm I haven't yet fully grasped your new proposal, but wanted to throw in a comment first:
Is there a problem to use your first implementation and consume all unsubscribe-messages on drop? (I really liked your first implementation and think it's quite ergonomic)

jwilm commented:
@badboy the initial implementation *could* leave the connection in an unusable state if it encounters an error while running the unsubscribe process. I made the changes proposed in my previous comment. It has the same API as the original proposal, but it ensures that we don't accidentally miss an error in the Drop impl and leave the connection in an unknown state.

badboy commented:
I will look into this soon.

jwilm commented:
No rush, take your time.

badboy commented on src/connection.rs:
```diff
@@ -427,44 +499,32 @@ impl ConnectionLike for Connection {
-        }
-        chan
+impl<'a> PubSub<'a> {
+    pub fn new(con: &'a mut Connection) -> Self {
```
Do we need to expose this constructor? I'd favor to only have `as_pubsub` on the connection.

jwilm replied:
Nope, this was a mistake. Nice catch.

badboy commented on src/connection.rs:
```diff
@@ -375,10 +371,82 @@ impl Connection {
     fn req_packed_command(&self, cmd: &[u8]) -> RedisResult<Value> {
+        if self.pubsub.get() {
+            self.exit_pubsub()?;
+        }
```
This now forces an additional check for each command sent, right? I guess the performance implications are still negligible?

jwilm replied:
It is an additional check, but I'm guessing the branch predictor will behave favorably here. In our benchmarking of this library at OneSignal, the parser ends up being a much bigger bottleneck than anything else (we have at least 1 PR forthcoming in that regard).

badboy replied:
Looking forward to that PR.

jwilm replied:
There's [a PR](https://github.com/mitsuhiko/redis-rs/pull/151) up for it as of a minute ago.

jwilm commented on src/commands.rs:
````diff
@@ -734,9 +734,108 @@ implement_commands! {
+/// # Ok(()) }
+/// ```
+//
+// In the future, it would be nice to implement Try such that `?` will work
````
I did not, but from what I understand about the Try trait, it should be possible.

jwilm commented:
Feedback addressed; can you take another look?

jwilm commented:
Note that I replied to one or two of the comments you left which have now been folded.

jwilm commented:
@badboy is there anything blocking this PR now?

badboy commented:
Just me hitting the big green button. Will do so today.
